payments/models.py

```diff
@@ -154,3 +169,356 @@ def __str__(self):
+        product = (
+            product
+            if product
+            else WinterStorageProduct.objects.filter(id=self._product_object_id).first()
+        )
+        # If for some reason neither was found (shouldn't be the case), raise an error
+        if not product:
+            raise ValidationError(_("The product passed is not valid"))
+
+        self.product = product
+
+    def _assign_lease_from_object_id(self):
+        # Try to get a BerthLease (BL)
+        lease = BerthLease.objects.filter(id=self._lease_object_id).first()
+        # If the BL was not found, try getting a WinterStorageLease
+        lease = (
+            lease
+            if lease
+            else WinterStorageLease.objects.filter(id=self._lease_object_id).first()
+        )
+        # If for some reason neither was found, raise an error
+        if not lease:
+            raise ValidationError(_("The lease passed is not valid"))
+
+        self.lease = lease
+
+    def save(self, *args, **kwargs):  # noqa: C901  is too complex
+        self.full_clean()
+
+        # If the product is being added from the admin (only the ID is passed)
+        if self._product_object_id and not self.product:
+            self._assign_product_from_object_id()
+
+        # If the lease is being added from the admin (only the ID is passed)
+        if self._lease_object_id and not self.lease:
+            self._assign_lease_from_object_id()
+
+        creating = self._state.adding
+        # If the product instance is being passed
+        # Price has to be assigned before saving but only if creating
+        if creating and self.product:
+            # Assign the price and tax from the product
+            if not self.price:
+                self.price = self.product.price_value
+            if not self.tax_percentage:
+                self.tax_percentage = self.product.tax_percentage
+
+        old_instance = Order.objects.filter(id=self.id).first()
+        # Save before adding additional products to have access to self
+        super().save(*args, **kwargs)
+
+        # Create OrderLines for the corresponding services only:
+        #   - If it's creating the order
+        #   - If the previous lease value was None (setting a lease)
+        # For now, it only applies BerthLeases, since WinterStorageLeases include the
+        # services on the m2 price.
+        if (creating or (old_instance and not old_instance.lease)) and isinstance(
+            self.lease, BerthLease
+        ):
+            self._create_order_lines(self.lease.berth.pier)
+
+
+class OrderLine(UUIDModel, TimeStampedModel):
+    order = models.ForeignKey(
+        Order,
+        verbose_name=_("order"),
+        related_name="order_lines",
+        on_delete=models.CASCADE,
+    )
+    product = models.ForeignKey(
+        AdditionalProduct,
+        verbose_name=_("product"),
+        related_name="orders",
+        on_delete=models.CASCADE,
```

Comment:
hm, what if at some point someone deletes the product? it would cascade this line, thus affecting the total order price and everything 🤔 maybe this should be nullable?

Reply:
True, it would make sense since we're keeping price and tax, so if the product would be deleted, we can still keep track.